src/ninja.cc

```diff
@@ -819,23 +825,45 @@ int NinjaMain::ToolInputs(const Options* options, int argc, char* argv[]) {
     return 1;
   }
 
-  InputsCollector collector;
-  for (const Node* node : nodes)
-    collector.VisitNode(node);
+  if (include_target) {
+    for (const Node* node : nodes) {
+      InputsCollector collector;
+
+      collector.VisitNode(node);
+      std::vector<std::string> inputs =
+          collector.GetInputsAsStrings(shell_escape);
 
-  std::vector<std::string> inputs = collector.GetInputsAsStrings(shell_escape);
-  if (!dependency_order)
-    std::sort(inputs.begin(), inputs.end());
+      if (!dependency_order)
+        std::sort(inputs.begin(), inputs.end());
 
-  if (print0) {
-    for (const std::string& input : inputs) {
-      fwrite(input.c_str(), input.size(), 1, stdout);
-      fputc('\0', stdout);
+      for (const std::string& input : inputs)
+        if (print0) {
+          printf("%s:%s", node->path().c_str(), input.c_str());
+          fputc('\0', stdout);
+        } else {
+          printf("%s:%s\n", node->path().c_str(), input.c_str());
+        }
     }
-    fflush(stdout);
   } else {
-    for (const std::string& input : inputs)
-      puts(input.c_str());
+    InputsCollector collector;
+    for (const Node* node : nodes)
+      collector.VisitNode(node);
+
+    std::vector<std::string> inputs =
+        collector.GetInputsAsStrings(shell_escape);
+    if (!dependency_order)
+      std::sort(inputs.begin(), inputs.end());
+
+    if (print0) {
+      for (const std::string& input : inputs) {
+        fwrite(input.c_str(), input.size(), 1, stdout);
+        fputc('\0', stdout);
+      }
+      fflush(stdout);
+    } else {
+      for (const std::string& input : inputs)
+        puts(input.c_str());
+    }
```

Comment:
Avoid code duplication by defining a local lambda that captures `shell_escape` and `dependency_order`, takes an `InputsCollector&` and ` const char* prefix` as input and does the printing. You'll be able to call it from the two branches of your `if (include_target)` condition.